File: sdks/go/escrow_test.go
package solvela

// Unit tests for the escrow deposit-transaction builder.
//
// The golden-vector parity test is the contract: the Go deposit builder MUST
// reproduce crates/escrow-tx/src/deposit.rs's GOLDEN_VECTOR_B64 byte-for-byte
// for its fixed input. If it diverges, the on-chain layout disagreement is a
// fund-misdirection bug — fix the Go builder, NEVER edit the expected golden
// value (the Rust/on-chain value is ground truth).

import (
	"crypto/ed25519"
	"encoding/base64"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"testing"

	"github.com/mr-tron/base58"
)

const (
	goldenProvider      = "9WzDXwBbmkg8ZTbNMqUxvQRAyrZzDsGYdLVL9zYtAWWM"
	goldenUSDCMint      = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"
	goldenEscrowProgram = "9neDHouXgEgHZDde5SpmqqEZ9Uv35hFcjtFEPxomtHLU"
)

// goldenKeypair builds the deterministic 64-byte agent keypair from seed
// [42]*32 (matches the Rust golden fixture: SigningKey::from_bytes([42; 32])).
func goldenKeypair() ed25519.PrivateKey {
	seed := make([]byte, 32)
	for i := range seed {
		seed[i] = 42
	}
	return ed25519.NewKeyFromSeed(seed)
}

func goldenParams() *DepositParams {
	var serviceID [32]byte
	for i := range serviceID {
		serviceID[i] = 7
	}
	var blockhash [32]byte
	for i := range blockhash {
		blockhash[i] = 0xAB
	}
	return &DepositParams{
		agentKeypair:       goldenKeypair(),
		ProviderWalletB58:  goldenProvider,
		USDCMintB58:        goldenUSDCMint,
		EscrowProgramIDB58: goldenEscrowProgram,
		Amount:             2625,
		ServiceID:          serviceID,
		ExpirySlot:         1_000_750,
		RecentBlockhash:    blockhash,
	}
}

// TestDepositTxMatchesGoldenVector is the sole correctness bar for the wire
// format: byte-for-byte parity with the canonical Rust builder.
func TestDepositTxMatchesGoldenVector(t *testing.T) {
	out, err := buildDepositTx(goldenParams())
	if err != nil {
		t.Fatalf("golden build should succeed: %v", err)
	}
	if out != GoldenVectorB64 {
		t.Fatalf("Go deposit-tx does not match the Rust golden vector.\n"+
			"The on-chain layout is ground truth — fix the Go builder, do NOT edit the expected value.\n"+
			"got:  %s\nwant: %s", out, GoldenVectorB64)
	}
}

// TestDepositTxDeterministic confirms the build is pure: same fixed input
// always yields the same bytes (no nonce, no clock).
func TestDepositTxDeterministic(t *testing.T) {
	a, err := buildDepositTx(goldenParams())
	if err != nil {
		t.Fatalf("build a: %v", err)
	}
	b, err := buildDepositTx(goldenParams())
	if err != nil {
		t.Fatalf("build b: %v", err)
	}
	if a != b {
		t.Fatal("deposit build is not deterministic")
	}
}

func TestGoldenVectorDecodesToSignedTx(t *testing.T) {
	raw, err := base64.StdEncoding.DecodeString(GoldenVectorB64)
	if err != nil {
		t.Fatalf("golden vector must be valid base64: %v", err)
	}
	// compact-u16(1) || sig(64) || message
	if raw[0] != 0x01 {
		t.Errorf("first byte should be 0x01 (1 signature), got 0x%02x", raw[0])
	}
	if len(raw) <= 1+ed25519.SignatureSize {
		t.Errorf("tx too short: %d bytes", len(raw))
	}
}

// TestGoldenVectorMatchesRustSource is the drift guard: the Go golden constant
// MUST equal the Rust source of truth. If the Rust GOLDEN_VECTOR_B64 ever
// changes, the on-chain wire layout changed and this test fails loudly, forcing
// a resync of the Go constant and builder rather than letting the SDKs silently
// diverge.
func TestGoldenVectorMatchesRustSource(t *testing.T) {
	// escrow_test.go -> sdks/go; worktree root is two levels up.
	root, err := filepath.Abs(filepath.Join("..", ".."))
	if err != nil {
		t.Fatalf("resolve worktree root: %v", err)
	}
	rustSrc := filepath.Join(root, "crates", "escrow-tx", "src", "deposit.rs")
	data, err := os.ReadFile(rustSrc)
	if err != nil {
		t.Fatalf("Rust deposit.rs not found at %s: %v", rustSrc, err)
	}
	re := regexp.MustCompile(`GOLDEN_VECTOR_B64:\s*&str\s*=\s*"([^"]+)"`)
	m := re.FindSubmatch(data)
	if m == nil {
		t.Fatalf("could not locate GOLDEN_VECTOR_B64 in %s", rustSrc)
	}
	rustValue := string(m[1])
	if rustValue != GoldenVectorB64 {
		t.Fatalf("Go GoldenVectorB64 has drifted from the Rust source of truth " +
			"(crates/escrow-tx/src/deposit.rs). The on-chain layout is ground truth: " +
			"resync the Go constant AND re-verify buildDepositTx byte-parity — do NOT " +
			"just paste the new value.")
	}
}

// ---------------------------------------------------------------------------
// Derivation parity (PDA seeds + vault ATA + discriminator).
// ---------------------------------------------------------------------------

func TestEscrowPDAExternalGroundTruth(t *testing.T) {
	// Mirrors crates/escrow-tx/src/pda.rs ground-truth test:
	// agent=[1]*32, service_id=[2]*32 -> BEAUsvsWvV4o6y7XkC1bkyTq4FtQnKErcV3dzTFPT5hX, bump 255.
	program, err := decodeBs58Pubkey(goldenEscrowProgram)
	if err != nil {
		t.Fatalf("decode program: %v", err)
	}
	var agent, serviceID [32]byte
	for i := range agent {
		agent[i] = 1
		serviceID[i] = 2
	}
	pda, bump, ok := findProgramAddress([][]byte{[]byte("escrow"), agent[:], serviceID[:]}, program)
	if !ok {
		t.Fatal("PDA derivation failed")
	}
	if got := base58.Encode(pda[:]); got != "BEAUsvsWvV4o6y7XkC1bkyTq4FtQnKErcV3dzTFPT5hX" {
		t.Errorf("PDA mismatch: got %s", got)
	}
	if bump != 255 {
		t.Errorf("bump mismatch: got %d, want 255", bump)
	}
}

func TestVaultATAKnownValue(t *testing.T) {
	// Mirrors crates/escrow-tx/src/pda.rs test_derive_ata_for_known_wallet.
	wallet, err := decodeBs58Pubkey("4P8mSmvv3nfzUtoqhNKG1mfGrHMVbXvKBXR7fDivv6qp")
	if err != nil {
		t.Fatalf("decode wallet: %v", err)
	}
	mint, err := decodeBs58Pubkey(goldenUSDCMint)
	if err != nil {
		t.Fatalf("decode mint: %v", err)
	}
	ata, ok := deriveATAAddress(wallet, mint)
	if !ok {
		t.Fatal("ATA derivation failed")
	}
	if got := base58.Encode(ata[:]); got != "CYHVCkLwiEjMBdRiz5MsrrCbVL2YTZuv57TjV3ggxoSN" {
		t.Errorf("ATA mismatch: got %s", got)
	}
}

func TestAnchorDiscriminatorDeterministic(t *testing.T) {
	d1 := anchorDiscriminator("deposit")
	d2 := anchorDiscriminator("deposit")
	if d1 != d2 {
		t.Error("discriminator not deterministic")
	}
	if anchorDiscriminator("deposit") == anchorDiscriminator("claim") {
		t.Error("distinct instruction names must give distinct discriminators")
	}
}

func TestIsOnEd25519CurveDetectsOffCurvePDA(t *testing.T) {
	// A derived PDA must be OFF the curve.
	program, _ := decodeBs58Pubkey(goldenEscrowProgram)
	var agent, serviceID [32]byte
	for i := range agent {
		agent[i] = 1
		serviceID[i] = 2
	}
	pda, _, _ := findProgramAddress([][]byte{[]byte("escrow"), agent[:], serviceID[:]}, program)
	if isOnEd25519Curve(pda) {
		t.Error("PDA must be off the ed25519 curve")
	}
	// A normal wallet pubkey (the golden agent) is ON the curve.
	var agentPub [32]byte
	copy(agentPub[:], goldenKeypair().Public().(ed25519.PublicKey))
	if !isOnEd25519Curve(agentPub) {
		t.Error("a real ed25519 pubkey must be on the curve")
	}
}

// ---------------------------------------------------------------------------
// Rejection / fail-closed behavior.
// ---------------------------------------------------------------------------

func TestDepositZeroAmountRejected(t *testing.T) {
	p := goldenParams()
	p.Amount = 0
	_, err := buildDepositTx(p)
	if err == nil {
		t.Fatal("zero amount should be rejected")
	}
	if _, ok := err.(*SignerError); !ok {
		t.Fatalf("expected *SignerError, got %T", err)
	}
}

func TestDepositInvalidKeypairLengthRejected(t *testing.T) {
	p := goldenParams()
	p.agentKeypair = ed25519.PrivateKey([]byte{1, 2, 3})
	_, err := buildDepositTx(p)
	if err == nil {
		t.Fatal("invalid keypair length should be rejected")
	}
}

func TestDepositInvalidProviderRejected(t *testing.T) {
	p := goldenParams()
	p.ProviderWalletB58 = "not-base58-0OIl!!!"
	_, err := buildDepositTx(p)
	if err == nil {
		t.Fatal("invalid provider address should be rejected")
	}
}

func TestDepositInvalidMintRejected(t *testing.T) {
	p := goldenParams()
	p.USDCMintB58 = "not-base58-0OIl!!!"
	_, err := buildDepositTx(p)
	if err == nil {
		t.Fatal("invalid mint address should be rejected")
	}
}

func TestDepositKeypairNotLeakedInError(t *testing.T) {
	// A malformed keypair must never echo secret bytes. We pass a wrong-length
	// key; the error message must not contain any hex/base58 of the bytes.
	p := goldenParams()
	p.agentKeypair = ed25519.PrivateKey([]byte{9, 9, 9, 9})
	_, err := buildDepositTx(p)
	if err == nil {
		t.Fatal("expected error")
	}
	if got := err.Error(); got != "signer error: invalid agent keypair: wrong length" {
		t.Errorf("error message leaked detail or changed: %q", got)
	}
}

func TestBuildLegacyMessageRejectsTooManyAccounts(t *testing.T) {
	// 127 data accounts + 1 program key = 128 total -> exceeds the limit.
	accounts := make([][32]byte, 127)
	_, err := buildLegacyMessage([3]byte{1, 0, 0}, accounts, [32]byte{}, [32]byte{}, 127, []byte{0}, []byte{0})
	if err == nil {
		t.Fatal("128 total accounts should be rejected")
	}
}

func TestBuildLegacyMessageRejectsLongIxData(t *testing.T) {
	_, err := buildLegacyMessage([3]byte{1, 0, 0}, [][32]byte{{}}, [32]byte{}, [32]byte{}, 1, []byte{0}, make([]byte, 128))
	if err == nil {
		t.Fatal("128 ix-data bytes should be rejected")
	}
}

// TestDepositCorruptedPubkeyHalfRejected mirrors the Rust guard
// (crates/escrow-tx/src/deposit.rs): ed25519.PrivateKey.Public() copies the
// stored bytes [32:64] rather than re-deriving from the seed, so a keypair with
// a corrupted public half would sign fine but seed the escrow PDA from the
// wrong identity, producing an un-claimable on-chain deposit. The builder must
// re-derive from the seed and reject the mismatch.
func TestDepositCorruptedPubkeyHalfRejected(t *testing.T) {
	p := goldenParams()
	corrupted := make(ed25519.PrivateKey, len(p.agentKeypair))
	copy(corrupted, p.agentKeypair)
	// Flip a bit in the stored public half (bytes 32..64) WITHOUT touching the
	// seed. The seed-derived pubkey will no longer match the stored one.
	corrupted[40] ^= 0xFF
	p.agentKeypair = corrupted

	_, err := buildDepositTx(p)
	if err == nil {
		t.Fatal("a keypair whose stored pubkey half does not match its seed must be rejected")
	}
	if _, ok := err.(*SignerError); !ok {
		t.Fatalf("expected *SignerError, got %T", err)
	}
	if !strings.Contains(err.Error(), "derived pubkey does not match stored pubkey") {
		t.Errorf("unexpected error: %q", err.Error())
	}
}

// TestDepositParamsRedactsKeypair verifies the secret keypair never appears in
// any fmt rendering of DepositParams (%v, %+v, %#v, %s). A raw 64-byte secret
// key landing in a log line is a key-disclosure bug.
func TestDepositParamsRedactsKeypair(t *testing.T) {
	// goldenParams() returns a *DepositParams; dereference so we can test both
	// the value and pointer forms below.
	pp := goldenParams()
	p := *pp
	// The first secret-seed byte of the golden keypair is 42 (0x2a). If the
	// struct were reflected normally, %v / %#v would print the keypair slice as
	// "[42 42 42 ...]" / "ed25519.PrivateKey{0x2a, ...}".
	for _, verb := range []string{"%v", "%+v", "%#v", "%s"} {
		// Test both the value and the *DepositParams pointer (production code
		// holds a *DepositParams and would log it directly).
		for _, arg := range []any{p, pp} {
			out := fmt.Sprintf(verb, arg)
			if !strings.Contains(out, "[REDACTED]") {
				t.Errorf("%s rendering of %T missing [REDACTED] marker: %q", verb, arg, out)
			}
			// Guard against the secret bytes leaking in either base-10 (reflected
			// slice) or 0x2a hex (%#v) form. "42 42 42" and "0x2a, 0x2a" are the
			// tell-tale shapes of a dumped 64-byte key.
			if strings.Contains(out, "42 42 42") || strings.Contains(out, "0x2a, 0x2a") {
				t.Errorf("%s rendering of %T leaked raw keypair bytes: %q", verb, arg, out)
			}
		}
	}
}

File: sdks/typescript/tests/unit/escrow.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';
import { Keypair } from '@solana/web3.js';

import {
  buildDepositTx,
  GOLDEN_VECTOR_B64,
  anchorDiscriminator,
  type DepositParams,
} from '../../src/escrow.js';
import { Wallet } from '../../src/wallet.js';
import { SignerError } from '../../src/errors.js';

// ---------------------------------------------------------------------------
// Golden-vector parity is the contract: the TS deposit builder MUST reproduce
// crates/escrow-tx/src/deposit.rs's GOLDEN_VECTOR_B64 byte-for-byte for its
// fixed input. If it diverges, the on-chain layout disagreement is a
// fund-misdirection bug — fix the TS builder, NEVER edit the expected golden
// value (the Rust/on-chain value is ground truth). Mirrors Go escrow_test.go +
// Python tests/unit/test_escrow.py.
// ---------------------------------------------------------------------------

const GOLDEN_PROVIDER = '9WzDXwBbmkg8ZTbNMqUxvQRAyrZzDsGYdLVL9zYtAWWM';
const GOLDEN_USDC_MINT = 'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v';
const GOLDEN_ESCROW_PROGRAM = '9neDHouXgEgHZDde5SpmqqEZ9Uv35hFcjtFEPxomtHLU';

/**
 * Deterministic agent wallet from seed [42; 32] — identical to the Rust/Go/
 * Python golden fixture (`SigningKey::from_bytes([42; 32])`).
 */
function goldenWallet(): Wallet {
  const kp = Keypair.fromSeed(new Uint8Array(32).fill(42));
  return Wallet.fromKeypairBytes(kp.secretKey);
}

function goldenParams(): DepositParams {
  return {
    wallet: goldenWallet(),
    providerWalletB58: GOLDEN_PROVIDER,
    usdcMintB58: GOLDEN_USDC_MINT,
    escrowProgramIdB58: GOLDEN_ESCROW_PROGRAM,
    amount: 2625,
    serviceId: new Uint8Array(32).fill(7),
    expirySlot: 1_000_750,
    recentBlockhash: new Uint8Array(32).fill(0xab),
  };
}

describe('Escrow deposit golden vector (byte-exact parity)', () => {
  it('reproduces the canonical Rust/Go golden vector byte-for-byte', () => {
    const out = buildDepositTx(goldenParams());
    // NEVER edit GOLDEN_VECTOR_B64 to match this output. The on-chain layout is
    // ground truth — if this diverges, the fix is in the builder.
    expect(out).toBe(GOLDEN_VECTOR_B64);
  });

  it('is deterministic for identical fixed input (no nonce, no clock)', () => {
    const a = buildDepositTx(goldenParams());
    const b = buildDepositTx(goldenParams());
    expect(a).toBe(b);
  });

  it('golden vector decodes to 0x01 || sig(64) || message', () => {
    const raw = Buffer.from(GOLDEN_VECTOR_B64, 'base64');
    // compact-u16(1) || sig(64) || message
    expect(raw[0]).toBe(0x01);
    expect(raw.length).toBeGreaterThan(1 + 64);
  });
});

describe('Escrow golden vector drift guard', () => {
  // The TS escrow golden constant MUST equal the Rust source of truth. If the
  // Rust GOLDEN_VECTOR_B64 ever changes, the on-chain wire layout changed and
  // this test fails loudly, forcing a resync rather than silent divergence.
  // Mirrors Go's TestGoldenVectorMatchesRustSource.
  it('TS escrow golden constant matches the Rust source of truth', () => {
    const here = dirname(fileURLToPath(import.meta.url));
    // tests/unit/escrow.test.ts -> worktree root is ../../../../ from here.
    const rustSrc = resolve(here, '../../../../crates/escrow-tx/src/deposit.rs');
    const text = readFileSync(rustSrc, 'utf-8');
    const m = text.match(/GOLDEN_VECTOR_B64:\s*&str\s*=\s*"([^"]+)"/);
    expect(m, 'could not locate GOLDEN_VECTOR_B64 in crates/escrow-tx/src/deposit.rs').not.toBeNull();
    const rustValue = m![1];
    expect(rustValue).toBe(GOLDEN_VECTOR_B64);
  });
});

describe('Escrow deposit derivation parity', () => {
  it('anchorDiscriminator is deterministic and distinct per instruction', () => {
    expect(anchorDiscriminator('deposit')).toEqual(anchorDiscriminator('deposit'));
    expect(anchorDiscriminator('deposit')).not.toEqual(anchorDiscriminator('claim'));
  });

  it('the deposit tx embeds the deposit discriminator', () => {
    const raw = Buffer.from(buildDepositTx(goldenParams()), 'base64');
    const disc = Buffer.from(anchorDiscriminator('deposit'));
    expect(raw.includes(disc)).toBe(true);
  });

  it('the deposit tx embeds the agent pubkey', () => {
    const params = goldenParams();
    const agentPubkey = Buffer.from(params.wallet.publicKey().toBytes());
    const raw = Buffer.from(buildDepositTx(params), 'base64');
    expect(raw.includes(agentPubkey)).toBe(true);
  });
});

describe('Escrow deposit fail-closed amount guards', () => {
  it('rejects zero amount', () => {
    expect(() => buildDepositTx({ ...goldenParams(), amount: 0 })).toThrow(SignerError);
  });

  it('rejects a negative amount', () => {
    expect(() => buildDepositTx({ ...goldenParams(), amount: -1 })).toThrow(/greater than zero|integer/);
  });

  it('rejects a non-integer (float) amount', () => {
    expect(() => buildDepositTx({ ...goldenParams(), amount: 2625.5 })).toThrow(/integer/);
  });

  it('rejects a NaN amount', () => {
    expect(() => buildDepositTx({ ...goldenParams(), amount: Number.NaN })).toThrow(/integer/);
  });

  it('rejects an amount above the safe-integer range', () => {
    expect(() =>
      buildDepositTx({ ...goldenParams(), amount: Number.MAX_SAFE_INTEGER + 2 }),
    ).toThrow(/safe-integer|integer/);
  });
});

describe('Escrow deposit fail-closed expiry-slot guards', () => {
  // expiry_slot is encoded directly into the on-chain u64. A 0/negative value is
  // a dead-on-arrival deposit; NaN/Infinity make u64LE's BigInt(...) throw a raw
  // untyped error. All must fail closed with a typed SignerError, parallel to
  // the amount guards. The production path always passes a valid positive slot,
  // so the golden vector is unaffected.
  it('rejects a zero expiry slot', () => {
    expect(() => buildDepositTx({ ...goldenParams(), expirySlot: 0 })).toThrow(SignerError);
  });

  it('rejects a negative expiry slot', () => {
    expect(() => buildDepositTx({ ...goldenParams(), expirySlot: -1 })).toThrow(
      /expiry_slot must be a positive integer/,
    );
  });

  it('rejects a non-integer (float) expiry slot', () => {
    expect(() => buildDepositTx({ ...goldenParams(), expirySlot: 1_000_750.5 })).toThrow(
      /expiry_slot must be a positive integer/,
    );
  });

  it('rejects a NaN expiry slot', () => {
    expect(() => buildDepositTx({ ...goldenParams(), expirySlot: Number.NaN })).toThrow(
      /expiry_slot must be a positive integer/,
    );
  });

  it('rejects an Infinity expiry slot', () => {
    expect(() =>
      buildDepositTx({ ...goldenParams(), expirySlot: Number.POSITIVE_INFINITY }),
    ).toThrow(/expiry_slot must be a positive integer/);
  });
});

describe('Escrow deposit fail-closed structural guards', () => {
  it('rejects a malformed provider address', () => {
    expect(() =>
      buildDepositTx({ ...goldenParams(), providerWalletB58: 'not-base58-0OIl!!!' }),
    ).toThrow(SignerError);
  });

  it('rejects a malformed mint address', () => {
    expect(() =>
      buildDepositTx({ ...goldenParams(), usdcMintB58: 'not-base58-0OIl!!!' }),
    ).toThrow(SignerError);
  });

  it('rejects a malformed escrow program id', () => {
    expect(() =>
      buildDepositTx({ ...goldenParams(), escrowProgramIdB58: 'not-base58-0OIl!!!' }),
    ).toThrow(SignerError);
  });

  it('rejects a service_id of the wrong length', () => {
    expect(() =>
      buildDepositTx({ ...goldenParams(), serviceId: new Uint8Array(31).fill(7) }),
    ).toThrow(/service_id/);
  });

  it('rejects a recent_blockhash of the wrong length', () => {
    expect(() =>
      buildDepositTx({ ...goldenParams(), recentBlockhash: new Uint8Array(31).fill(0xab) }),
    ).toThrow(/blockhash/);
  });
});

describe('Escrow deposit corrupted-pubkey-half guard', () => {
  // The canonical Rust/Go builders re-derive the pubkey from the seed and reject
  // a keypair whose stored public half (bytes 32..64) does not match — a corrupt
  // half would sign fine but seed the escrow PDA from the wrong identity,
  // producing an un-claimable deposit. In the TS SDK this invariant lives at the
  // layer that actually enforces it: `Wallet.fromKeypairBytes` -> web3.js
  // `Keypair.fromSecretKey` validates the public half at construction and
  // rejects a corrupted keypair before it can ever reach the builder. Because a
  // `DepositParams.wallet` is, by type, an already-constructed `Wallet`, the
  // builder cannot receive a public-half-inconsistent keypair and therefore no
  // longer re-checks it (unlike Go, whose `DepositParams` takes raw
  // `ed25519.PrivateKey` bytes with no construction-time validation, and unlike
  // a builder-level guard that would have to pull raw secret-key material onto
  // the per-call money path for zero marginal safety). We pin the protection at
  // the Wallet construction boundary.
  it('rejects a keypair whose stored pubkey half does not match its seed (at the Wallet boundary)', () => {
    const kp = Keypair.fromSeed(new Uint8Array(32).fill(42));
    const corrupted = Uint8Array.from(kp.secretKey);
    // Flip a bit in the stored public half (bytes 32..64) WITHOUT touching the
    // seed (bytes 0..32).
    corrupted[40] ^= 0xff;
    expect(() => Wallet.fromKeypairBytes(corrupted)).toThrow();
  });
});

describe('Escrow deposit secret-redaction', () => {
  // The agent secret key material must never appear in any error message. We
  // force a builder error (malformed provider address) using the golden wallet
  // whose seed bytes are all 42 (0x2a) and assert the raw seed bytes never leak
  // into the thrown error.
  it('never leaks raw keypair bytes in a builder error message', () => {
    try {
      buildDepositTx({ ...goldenParams(), providerWalletB58: 'not-base58-0OIl!!!' });
      throw new Error('expected build to reject the malformed provider address');
    } catch (e) {
      const msg = e instanceof Error ? e.message : String(e);
      // 42 42 42 (decimal) / 2a,2a,2a (hex) are the tell-tale shapes of a dumped
      // 32-byte seed. Neither may appear.
      expect(msg).not.toMatch(/42, ?42, ?42/);
      expect(msg).not.toMatch(/2a,? ?2a,? ?2a/i);
    }
  });
});
